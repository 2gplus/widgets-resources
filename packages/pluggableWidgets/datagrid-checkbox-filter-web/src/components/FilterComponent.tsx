import { CSSProperties, ReactElement, useCallback, useEffect, useRef, useState,createElement } from "react";
import { debounce } from "@mendix/piw-utils-internal";

import classNames from "classnames";
import { DefaultFilterEnum } from "../../typings/DatagridCheckBoxFilterProps";

interface FilterComponentProps
{
    className?: string;
    defaultFilter: DefaultFilterEnum;
    delay: number;
    id?: string;
    placeholder?: string;
    screenReaderButtonCaption?: string;
    screenReaderInputCaption?: string;
    tabIndex?: number;
    styles?: CSSProperties;
    updateFilters?: (value: boolean | undefined, type: DefaultFilterEnum) => void;
    value?: boolean;
}

export function FilterComponent(props: FilterComponentProps): ReactElement
{
    const [type] = useState<DefaultFilterEnum>(props.defaultFilter);
    const [value, setValue] = useState<boolean | undefined>(undefined);
    const [valueInput, setValueInput] = useState<Number>(0);
    const inputRef = useRef<HTMLInputElement | null>(null);

    useEffect(() =>
    {
        if (props.value)
        {
            setValueInput(props.value ? 1 : 0);
            setValue(props.value);
        }
    }, [props.value]);

    useEffect(() =>
    {
        props.updateFilters?.(value, type);
    }, [value, type]);

    const onChange = useCallback(
        debounce((value?: boolean) => setValue(value), props.delay),
        [props.delay]
    );


    return (
        <div
            className={classNames("filter-container", props.className)}
            data-focusindex={props.tabIndex ?? 0}
            style={props.styles}
        >
            <input
                type={'checkbox'}
                aria-label={props.screenReaderInputCaption}
                onChange={e =>
                {
                    const value = e.target.checked;
                    setValueInput(value ? 1: 0);
                    onChange(value);
                }}
                placeholder={props.placeholder}
                ref={inputRef}
                checked={valueInput === 1}
            />
        </div>
    );
}
