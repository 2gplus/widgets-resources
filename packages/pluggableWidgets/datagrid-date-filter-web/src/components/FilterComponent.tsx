import { createElement, Dispatch, ReactElement, useCallback, useEffect, useRef, useState } from "react";
import { FilterSelector } from "./FilterSelector";
import { ListAttributeValue } from "mendix";
import { FilterCondition } from "mendix/filters";
import {
    and,
    attribute,
    greaterThan,
    greaterThanOrEqual,
    lessThan,
    lessThanOrEqual,
    literal
} from "mendix/filters/builders";
import { DefaultFilterEnum } from "../../typings/DatagridDateFilterProps";

import { changeTimeOfDate } from "../utils/utils";
import DatePickerComponent from "react-datepicker";
import { DatePicker } from "./DatePicker";

interface FilterComponentProps {
    adjustable: boolean;
    defaultFilter: DefaultFilterEnum;
    defaultValue?: Date;
    dateFormat?: string;
    filterDispatcher: Dispatch<{ getFilterCondition(attribute: ListAttributeValue): FilterCondition | undefined }>;
    locale?: string;
    name?: string;
    placeholder?: string;
    screenReaderButtonCaption?: string;
    screenReaderInputCaption?: string;
    tabIndex?: number;
}

export function FilterComponent(props: FilterComponentProps): ReactElement {
    const [type, setType] = useState<DefaultFilterEnum>(props.defaultFilter);
    const [value, setValue] = useState<Date | null>(null);
    const pickerRef = useRef<DatePickerComponent | null>(null);

    useEffect(() => {
        if (props.defaultValue) {
            setValue(props.defaultValue);
        }
    }, [props.defaultValue]);

    useEffect(() => {
        props.filterDispatcher({
            getFilterCondition: attr => {
                if (!attr.filterable || !value) {
                    return undefined;
                }

                const filterAttribute = attribute(attr.id);

                switch (type) {
                    case "greater":
                        return greaterThan(filterAttribute, literal(changeTimeOfDate(value, 23, 59, 59)));
                    case "greaterEqual":
                        return greaterThanOrEqual(filterAttribute, literal(changeTimeOfDate(value, 0, 0, 0)));
                    case "equal":
                        return and(
                            greaterThanOrEqual(filterAttribute, literal(changeTimeOfDate(value, 0, 0, 0))),
                            lessThanOrEqual(filterAttribute, literal(changeTimeOfDate(value, 23, 59, 59)))
                        );
                    case "notEqual":
                        return and(
                            lessThan(filterAttribute, literal(changeTimeOfDate(value, 0, 0, 0))),
                            greaterThan(filterAttribute, literal(changeTimeOfDate(value, 23, 59, 59)))
                        );
                    case "smaller":
                        return lessThan(filterAttribute, literal(changeTimeOfDate(value, 0, 0, 0)));
                    case "smallerEqual":
                        return lessThanOrEqual(filterAttribute, literal(changeTimeOfDate(value, 23, 59, 59)));
                }
            }
        });
    }, [props.filterDispatcher, value, type]);

    const focusInput = useCallback(() => {
        if (pickerRef.current) {
            pickerRef.current.setFocus();
        }
    }, [pickerRef.current]);

    return (
        <div className="filter-container" data-focusindex={props.tabIndex ?? 0}>
            {props.adjustable && (
                <FilterSelector
                    ariaLabel={props.screenReaderButtonCaption}
                    defaultFilter={props.defaultFilter}
                    name={props.name}
                    onChange={type => {
                        setType(type);
                        focusInput();
                    }}
                />
            )}
            <DatePicker
                adjustable={props.adjustable}
                dateFormat={props.dateFormat}
                locale={props.locale}
                name={props.name}
                placeholder={props.placeholder}
                ref={pickerRef}
                screenReaderInputCaption={props.screenReaderInputCaption}
                setValue={setValue}
                value={value}
            />
        </div>
    );
}
