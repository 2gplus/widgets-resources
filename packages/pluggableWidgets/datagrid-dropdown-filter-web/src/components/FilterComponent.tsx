import { createElement, Dispatch, Fragment, ReactElement, useCallback, useEffect, useRef, useState } from "react";
import { ListAttributeValue, ObjectItem } from "mendix";
import { useOnClickOutside } from "@widgets-resources/piw-utils";

interface Option {
    caption: string;
    value: string;
}

interface FilterComponentProps {
    ariaLabel?: string;
    emptyOptionCaption?: string;
    filterDispatcher: Dispatch<{ filter(item: ObjectItem, attribute: ListAttributeValue): boolean }>;
    multiSelect?: boolean;
    name?: string;
    options: Option[];
    tabIndex?: number;
    defaultValue?: string;
}

export function FilterComponent(props: FilterComponentProps): ReactElement {
    const [valueInput, setValueInput] = useState("");
    const [options, setOptions] = useState<Option[]>([]);
    const [selectedFilters, setSelectedFilters] = useState<Option[]>([]);
    const [show, setShow] = useState(false);
    const [dropdownWidth, setDropdownWidth] = useState(0);

    const listRef = useRef<HTMLUListElement>(null);

    const setMultiSelectFilters = useCallback(
        (selectedOptions: Option[]) => {
            if (selectedOptions?.length === 0) {
                setValueInput(props.emptyOptionCaption ?? "");
                setSelectedFilters([]);
            } else {
                setValueInput(selectedOptions.map(option => option.caption).join(","));
                setSelectedFilters(selectedOptions);
            }
        },
        [props.emptyOptionCaption]
    );

    const onClick = useCallback(
        (option: Option) => {
            if (props.multiSelect) {
                setMultiSelectFilters(toggleFilter(selectedFilters, option));
            } else {
                setValueInput(option.caption);
                setSelectedFilters(option.value ? [option] : []);
                setShow(false);
            }
        },
        [selectedFilters, props.emptyOptionCaption, props.multiSelect]
    );

    useOnClickOutside(listRef, () => setShow(false));

    // Select the first option Or default option on load
    useEffect(() => {
        if (props.multiSelect) {
            if (props.defaultValue) {
                const initialOptions = props.defaultValue
                    .split(",")
                    .map(value => props.options.find(option => option.value === value))
                    .filter(Boolean) as Option[];

                // User can set anything, but it could not match so we have to set to empty or ""
                setMultiSelectFilters(initialOptions);
            } else {
                setValueInput(props.emptyOptionCaption ?? "");
            }

            setOptions(props.options);
        } else {
            // We want to add empty option caption
            const optionsWithEmptyCaption = [{ caption: props.emptyOptionCaption ?? "", value: "" }, ...props.options];
            const initialOption =
                optionsWithEmptyCaption.find(option => option.value === props.defaultValue) ??
                optionsWithEmptyCaption[0];

            setValueInput(initialOption?.caption ?? "");
            setSelectedFilters([initialOption]);
            setOptions(optionsWithEmptyCaption);
        }
    }, [props.defaultValue, props.options, props.emptyOptionCaption]);

    // Filter
    useEffect(() => {
        if (props.filterDispatcher) {
            props.filterDispatcher({
                filter: (item, attr): boolean => {
                    if (selectedFilters.length > 0) {
                        return selectedFilters.some(
                            selectedFilter =>
                                attr(item).value?.toString().toLocaleLowerCase() ===
                                    selectedFilter.value?.toString().toLocaleLowerCase() ||
                                selectedFilter.value?.toString() === ""
                        );
                    }
                    return true;
                }
            });
        }
    }, [props.filterDispatcher, selectedFilters]);

    return (
        <div className="dropdown-container" data-focusindex={props.tabIndex ?? 0}>
            <input
                value={valueInput}
                className="form-control dropdown-triggerer"
                onClick={() => setShow(true)}
                onFocus={() => setShow(true)}
                aria-haspopup
                ref={inputRef => {
                    if (inputRef && inputRef.clientWidth) {
                        setDropdownWidth(inputRef.clientWidth);
                    }
                }}
                aria-expanded={show}
                aria-controls={`${props.name}-dropdown-list`}
                aria-label={props.ariaLabel}
            />
            <ul
                id={`${props.name}-dropdown-list`}
                className={`dropdown-list ${show && "dropdown-list-visible"}`}
                ref={listRef}
                style={{ width: dropdownWidth }}
                role="menu"
                data-focusindex={0}
            >
                {options.map((option, index) => (
                    <li
                        className={!props.multiSelect && selectedFilters.includes(option) ? "filter-selected" : ""}
                        key={index}
                        onClick={() => onClick(option)}
                        onKeyDown={e => {
                            if (e.key === "Enter" || e.key === " ") {
                                e.preventDefault();
                                onClick(option);
                            }
                        }}
                        role="menuitem"
                        tabIndex={0}
                    >
                        {props.multiSelect ? (
                            <Fragment>
                                <input
                                    id={`checkbox_toggle_${index}`}
                                    type="checkbox"
                                    checked={selectedFilters.includes(option)}
                                />
                                <label htmlFor={`checkbox_toggle_${index}`}>{option.caption}</label>
                            </Fragment>
                        ) : (
                            <div className="filter-label">{option.caption}</div>
                        )}
                    </li>
                ))}
            </ul>
        </div>
    );
}

function toggleFilter(filters: Option[], filterToToggle: Option): Option[] {
    const alteredFilters = [...filters];
    const index = filters.indexOf(filterToToggle);
    if (index > -1) {
        alteredFilters.splice(index, 1);
    } else {
        alteredFilters.push(filterToToggle);
    }

    return alteredFilters;
}
