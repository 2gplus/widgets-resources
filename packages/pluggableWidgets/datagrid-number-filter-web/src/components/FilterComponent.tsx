import { createElement, Dispatch, ReactElement, useCallback, useEffect, useRef, useState } from "react";
import { FilterSelector } from "./FilterSelector";
import { ListAttributeValue } from "mendix";
import {
    attribute,
    equals,
    greaterThan,
    greaterThanOrEqual,
    lessThan,
    lessThanOrEqual,
    literal,
    notEqual
} from "mendix/filters/builders";
import { DefaultFilterEnum } from "../../typings/DatagridNumberFilterProps";
import { debounce } from "../utils/utils";
import { Big } from "big.js";
import classNames from "classnames";
import { FilterFunction } from "../utils/provider";
import { Alert } from "@mendix/piw-utils-internal";

function getAttributeTypeErrorMessage(type?: string): string | null {
    return type && !type.match(/AutoNumber|Decimal|Integer|Long/)
        ? "The attribute type being used for Data grid number filter is not 'Auto number, Decimal, Integer or Long'"
        : null;
}

interface FilterComponentProps {
    adjustable: boolean;
    attribute?: ListAttributeValue;
    defaultFilter: DefaultFilterEnum;
    delay: number;
    filterDispatcher: Dispatch<FilterFunction>;
    name?: string;
    placeholder?: string;
    screenReaderButtonCaption?: string;
    screenReaderInputCaption?: string;
    tabIndex?: number;
    value?: Big;
}

export function FilterComponent(props: FilterComponentProps): ReactElement {
    const [type, setType] = useState<DefaultFilterEnum>(props.defaultFilter);
    const [value, setValue] = useState<Big | undefined>(undefined);
    const [valueInput, setValueInput] = useState<string | undefined>(undefined);
    const inputRef = useRef<HTMLInputElement | null>(null);

    useEffect(() => {
        if (props.value) {
            setValueInput(props.value.toString());
            setValue(props.value);
        }
    }, [props.value]);

    useEffect(() => {
        if (props.filterDispatcher) {
            props.filterDispatcher({
                getFilterCondition: () => {
                    if (!props.attribute || !props.attribute.filterable || !value) {
                        return undefined;
                    }
                    const filterAttribute = attribute(props.attribute.id);
                    const filterValue = new Big(value);

                    switch (type) {
                        case "greater":
                            return greaterThan(filterAttribute, literal(filterValue));
                        case "greaterEqual":
                            return greaterThanOrEqual(filterAttribute, literal(filterValue));
                        case "equal":
                            return equals(filterAttribute, literal(filterValue));
                        case "notEqual":
                            return notEqual(filterAttribute, literal(filterValue));
                        case "smaller":
                            return lessThan(filterAttribute, literal(filterValue));
                        case "smallerEqual":
                            return lessThanOrEqual(filterAttribute, literal(filterValue));
                    }
                }
            });
        }
    }, [props.attribute, props.filterDispatcher, value, type]);

    const onChange = useCallback(
        debounce((value?: Big) => setValue(value), props.delay),
        [props.delay]
    );

    const focusInput = useCallback(() => {
        if (inputRef.current) {
            inputRef.current.focus();
        }
    }, [inputRef]);

    const errorMessage = getAttributeTypeErrorMessage(props.attribute?.type);

    if (errorMessage) {
        return <Alert bootstrapStyle="danger">{errorMessage}</Alert>;
    }

    return (
        <div className="filter-container" data-focusindex={props.tabIndex ?? 0}>
            {props.adjustable && (
                <FilterSelector
                    ariaLabel={props.screenReaderButtonCaption}
                    name={props.name}
                    defaultFilter={props.defaultFilter}
                    onChange={type => {
                        setType(type);
                        focusInput();
                    }}
                />
            )}
            <input
                aria-label={props.screenReaderInputCaption}
                className={classNames("form-control", { "filter-input": props.adjustable })}
                onChange={e => {
                    const value = e.target.value;
                    if (value && !isNaN(Number(value))) {
                        setValueInput(value);
                        onChange(new Big(Number(value)));
                    } else {
                        setValueInput(value);
                        onChange(undefined);
                    }
                }}
                placeholder={props.placeholder}
                ref={inputRef}
                type="number"
                value={valueInput}
            />
        </div>
    );
}
