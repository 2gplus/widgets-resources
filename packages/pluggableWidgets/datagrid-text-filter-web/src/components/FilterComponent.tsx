import { createElement, Dispatch, ReactElement, useCallback, useEffect, useState } from "react";
import { FilterSelector } from "./FilterSelector";
import { ObjectItem, ListAttributeValue } from "mendix";
import { DefaultFilterEnum } from "../../typings/DatagridTextFilterProps";
import { debounce } from "../utils/utils";

interface FilterComponentProps {
    ariaLabel?: string;
    defaultFilter: DefaultFilterEnum;
    delay: number;
    filterDispatcher: Dispatch<{ filter(item: ObjectItem, attribute: ListAttributeValue): boolean }>;
    name?: string;
    placeholder?: string;
    tabIndex?: number;
    value?: string;
}

export function FilterComponent(props: FilterComponentProps): ReactElement {
    const [type, setType] = useState<DefaultFilterEnum>(props.defaultFilter);
    const [value, setValue] = useState("");
    const [valueInput, setValueInput] = useState("");

    useEffect(() => {
        if (props.value) {
            setValueInput(props.value);
            setValue(props.value);
        }
    }, [props.value]);

    useEffect(() => {
        if (props.filterDispatcher) {
            props.filterDispatcher({
                filter: (item, attr): boolean => {
                    if (!value) {
                        return true;
                    }
                    const dataValue = attr(item).displayValue.toLowerCase();
                    const filterValue = value.toLowerCase();
                    switch (type) {
                        case "contains":
                            return dataValue.includes(filterValue);
                        case "startsWith":
                            return dataValue.startsWith(filterValue);
                        case "endsWith":
                            return dataValue.endsWith(filterValue);
                        case "greater":
                            return dataValue > filterValue;
                        case "greaterEqual":
                            return dataValue >= filterValue;
                        case "equal":
                            return dataValue === filterValue;
                        case "notEqual":
                            return dataValue !== filterValue;
                        case "smaller":
                            return dataValue < filterValue;
                        case "smallerEqual":
                            return dataValue <= filterValue;
                    }
                }
            });
        }
    }, [props.filterDispatcher, value, type]);

    const onChange = useCallback(
        debounce((value: string) => setValue(value), props.delay),
        [props.delay]
    );

    return (
        <div className="filter-container" data-focusindex={props.tabIndex ?? 0}>
            <FilterSelector name={props.name} defaultFilter={props.defaultFilter} onChange={setType} />
            <input
                aria-label={props.ariaLabel}
                className="form-control filter-input"
                onChange={e => {
                    setValueInput(e.target.value);
                    onChange(e.target.value);
                }}
                placeholder={props.placeholder}
                type="text"
                value={valueInput}
            />
        </div>
    );
}
