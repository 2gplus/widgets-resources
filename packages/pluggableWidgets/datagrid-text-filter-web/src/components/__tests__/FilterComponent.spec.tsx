import { shallow } from "enzyme";
import { createElement } from "react";
import { FilterComponent } from "../FilterComponent";

describe("Filter selector", () => {
    it("renders correctly", () => {
        const component = shallow(<FilterComponent filterDispatcher={jest.fn()} />);

        expect(component).toMatchSnapshot();
    });

    it("calls filterDispatcher when value changes", () => {
        const filterDispatcher = jest.fn();
        const component = shallow(<FilterComponent filterDispatcher={filterDispatcher} />);

        const input = component.find("input");
        input.simulate("change", { target: { value: "test" } });

        expect(filterDispatcher).toBeCalled();
    });
});
