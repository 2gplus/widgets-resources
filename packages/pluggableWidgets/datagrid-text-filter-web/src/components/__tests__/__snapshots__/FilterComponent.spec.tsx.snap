// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter selector renders correctly 1`] = `
<div
  className="filter-container"
>
  <FilterSelector
    defaultFilter="contains"
    onChange={[Function]}
  />
  <input
    className="form-control"
    onChange={[Function]}
    type="text"
  />
</div>
`;
